### DISCUSSION TITLE
[LUA] Can you unsubscribe from register_frame, register_start etc?

### DISCUSSION THREAD

**Discussion by Jakobud** [mamedev/discussions#158] [Q&A] [answered]
For older versions of MAME that still rely on `register_frame`, `register_start`, etc notifiers, is there some sort of "unsubscribe" functionality for them?

**Comment by cuavas** [comment, marked as answer]
`register_start`, `register_frame`, etc. are deprecated and will be removed at some point.  You should use `add_machine_reset_notifier`, `add_machine_frame_notifier`, etc. (see the [Emulator Interface](https://docs.mamedev.org/luascript/ref-common.html#emulator-interface) documentation).  These functions return [notifier subscriptions](https://docs.mamedev.org/luascript/ref-core.html#notifier-subscription) which allow unsubscribing.

You can’t manage subscriptions created with the `register_X` functions.  The closest you can do is to subscribe using an anonymous trampoline function that calls the real function by name, and then reassign the name with a function that does nothing (or does something else) when you no longer need it.

For example:
```lua
function on_frame()
    -- do something here
end

emu.register_frame(function () on_frame() end)

-- replace frame function with one that does nothing
on_frame = function () end
```

Note that all the callbacks will still be called, so it’s not a great idea to try and subscribe and unsubscribe numerous times like this.

**Reply by Jakobud** [reply]
That was the approach I was assuming would be best, but yes I agree not ideal. I am aware of the newer notifier functions. I was just writing a plugin that would be backwards compatible with the register_ functions. Thank you,

**Reply by Jakobud** [reply]
@cuavas is there a replacement for `register_frame_done`?

I'm trying to understand if you are going to draw anything on the screen it should be done in `register_frame_done` instead of `add_machine_frame_notifier` right?

**Reply by cuavas** [reply]
They’re called at different times:
* `add_machine_frame_notifier` functions are called at the end of each frame for the first emulated screen in the system.  This is when MAME synchronises with the wall clock if throttling is enabled, etc.
* `register_frame_done` functions are called for every frame update.  This happens at the end of frames for the first emulated screen, and also when triggered by the debugger, etc.  This is when inputs are updates, video output is updated if the frame isn’t skipped, the internal UI is updated, etc.

They’re for different things – one is for doing something for each emulated frame, the other is for doing something for each update.  The included `timecode` plugin does different things in each of those situations, for example.

**Reply by Jakobud** [reply]
@cuavas So I've got some code like this in my plugin

```
startNotifier = emu.add_machine_reset_notifier(function()

      local rom = emu.romname()

      -- If no rom is loaded, don't do anything
      if rom == "___empty" then
        return
      end

      local screen = manager.machine.screens[':screen']
      local centerX = screen.width / 2
      local centerY = screen.height / 2

      function draw_overlay()
        screen:draw_text(centerX, centerY, rom, 0xffffffff, 0xff000000) -- (x0, y0, msg)
      end

      emu.register_frame_done(draw_overlay, 'frame')

end)
```

It just writes the name of the rom in the middle of the screen when the game is running. Seems to work fine. But when I exit the game back to the MAME menu ui I get this

```
-----------------------------------------------------
Exception at EIP=00007ff6845ebbe8 (render_manager::ui_aspect(render_container*)+0x0108): ACCESS VIOLATION
While attempting to read memory at ffffffffffffffff
-----------------------------------------------------
RAX=0000000000000000 RBX=0000000000000002 RCX=00ffffff00ffffff RDX=0000000000000000
RSI=0000000000000000 RDI=0000000000000000 RBP=000001a6ccea1d88 RSP=00000041624f8c98
 R8=00000041624f8e60  R9=7efefefeff316a6c R10=000001a6cd7214d0 R11=8101010101010100
R12=00ffffff00ffffff R13=0000000000000003 R14=000001a6cd7f24b0 R15=0000000000000101
-----------------------------------------------------
Stack crawl:
  00000041624f8c90: 00007ff6845ebbe8 (render_manager::ui_aspect(render_container*)+0x0108)
  00000041624f8dd0: 00007ff68af97dc1 (mame_ui_manager::draw_text_full(render_container&, std::basic_string_view<char, std::char_traits<char> >, float, float, float, ui::text_layout::text_justify, ui::text_layout::word_wrapping, mame_ui_manager::draw_mode, rgb_t, rgb_t, float*, float*)+0x0101)
  00000041624f8f10: 00007ff68ae4bd81 (decltype(auto) sol::stack::stack_detail::eval<false, float, char const*, std::optional<unsigned int>, std::optional<unsigned int>, 2ull, 3ull, 4ull, 5ull, sol::argument_handler<sol::types<void, screen_device&, sol::basic_object<sol::basic_reference<false> >, float, char const*, std::optional<unsigned int>, std::optional<unsigned int> > >&, sol::wrapper<lua_engine::initialize()::{lambda(screen_device&, sol::basic_object<sol::basic_reference<false> >, float, char const*, std::optional<unsigned int>, std::optional<unsigned int>)#109}, void>::caller, lua_engine::initialize()::{lambda(screen_device&, sol::basic_object<sol::basic_reference<false> >, float, char const*, std::optional<unsigned int>, std::optional<unsigned int>)#109}&, screen_device&, sol::basic_object<sol::basic_reference<false> > >(sol::types<float, char const*, std::optional<unsigned int>, std::optional<unsigned int> >, std::integer_sequence<unsigned long long, 2ull, 3ull, 4ul+0x0341)
...
```

If I comment out the line to `draw_text` then I don't get this memory issue when exiting the game. Is the LUA engine still trying to write my rom name to the middle of the screen even though I'm not in a game? Is the solution to this to use the trampoline function you described to enable/disable the writing to the screen using `add_machine_reset_notifier` and ` add_machine_stop_notifier` ?

**Reply by Jakobud** [reply]
Okay I tried the trampoline function approach and it seemed to work, however, it looks like `emu.register_frame_done` will continue running even when I exit a game and pick a different game. Then both event handlers will be running and writing to the screen. Is there not an unsubscribe option for `register_frame_done`? What option do I have here? I guess declaring `register_frame_done` only once outside of `add_machine_reset_notifier` is the only option I can see at the moment.

**Reply by cuavas** [reply]
You don’t want to keep registering additional handlers, and you want to clean up anything tied to the machine when it exits.

The way you have that code, you'd register an additional handler each time the user does an F3 reset even if they never start another machine.

**Reply by Jakobud** [reply]
Yes I have finally realized that thank you. 

Just to verify, there is no way to unsubscribe from the `emu.register_frame_done` function correct?

**Reply by cuavas** [reply]
> Just to verify, there is no way to unsubscribe from the `emu.register_frame_done` function correct?

Correct, one of the major limitations of those `emu.register_xxx` functions is that you can’t unsubscribe.

**Reply by Jakobud** [reply]
When you use `emu.register_start` when you load up a game then exit and load another game, etc.. are you just creating more and more event handlers that are running in the background?

I'm just writing a plugin that it would be nice if it was a little bit backwards compatible so I'm curious.
